README.md
**Summary**: Secure SAML Service Provider

**Description**: Secure SAML Service Provider with a focus on saml2int 
compatibility and easy integration in PHP applications.

**License**: MIT

# Introduction

A SAML Service Provider (SP) with an easy API to use SAML authentication from
your existing PHP applications.

This project received a comprehensive security audit by 
[Cure53](https://cure53.de/) in November 2020. The identified issues were 
fixed. You can read the audit [report](audit/DEC-01-report.final.pdf) 
yourself.

# Why

There are various options for integrating SAML in your PHP application. 
However, most are either (very) complicated, include too many (useless) 
features, have hard requirements on [Apache](https://httpd.apache.org/) and are 
not easy to package for server operating systems like CentOS/Fedora and/or 
Debian.

We only need SAML SP support, so there is no need to include any IdP features, 
or other (obsolete) authentication protocols.

In addition, we only implement what is actually used "in the field" *and* that 
which is secure. So you won't find SHA1 support or insecure encryption.

# Features

- Only SAML SP functionality
- Only HTTP-Redirect for sending `AuthnRequest`, `LogoutRequest` to IdP
- Only HTTP-Redirect binding for receiving `LogoutResponse` from IdP
- Only HTTP-POST binding for receiving `Response` from IdP
- Always signs `AuthnRequest`
- Always signs `LogoutRequest`
- Supports signed `samlp:Response` and/or signed 
  `samlp:Response/saml:Assertion`
- Supports multiple IdP certificates for key rollover
- Allow specifying `AuthnContextClassRef` as part of the `AuthnRequest`
- Serializes `eduPersonTargetedID` as `idpEntityId!spEntityId!persistentId` 
  (like Shibboleth);
- Only supports `urn:oid` SAML attributes from a list of allowed attributes, 
  ignores the rest
- Verify "scope" of attributes based on `<shibmd:Scope>` metadata element when
  the IdP metadata contains this element
    - Silently removes the attribute (value) when scope does not match
- Converts `urn:oid` attribute names to "friendly" names for use by 
  applications
- Supports automated metadata retrieval/update with XML schema / signature 
  verification, see [METADATA](METADATA.md)
- Validates XML schema(s) when processing XML protocol messages
- Tested with IdPs:
    - [simpleSAMLphp](https://simplesamlphp.org/)
    - [OpenConext](https://openconext.org/)
    - [FrkoIdP](https://github.com/fkooman/php-saml-idp/)
    - [AD FS](https://en.wikipedia.org/wiki/Active_Directory_Federation_Services)
    - [Shibboleth IdP](https://www.shibboleth.net/products/identity-provider/)
- Simple built-in WAYF when more than 1 IdP is configured for the SP
- Support external discovery services implementing 
  [Identity Provider Discovery Service Protocol and Profile](https://docs.oasis-open.org/security/saml/Post2.0/sstc-saml-idp-discovery.html)

## SAML V2.0 Deployment Profile for Federation Interoperability 

We _do_ aim to eventually support everything as mentioned in 
[SAML V2.0 Deployment Profile for Federation Interoperability](https://kantarainitiative.github.io/SAMLprofiles/saml2int.html).

### Cryptographic Algorithms

| Type                 | Algorithm(s)                                          | Supported | 
| -------------------- | ----------------------------------------------------- | --------- |
| Digest               | `http://www.w3.org/2001/04/xmlenc#sha256`             | Yes       |
| Signature            | `http://www.w3.org/2001/04/xmldsig-more#rsa-sha256`   | Yes       |
|                      | `http://www.w3.org/2001/04/xmldsig-more#ecdsa-sha256` | No        |
| Block Encryption     | `http://www.w3.org/2009/xmlenc11#aes128-gcm`          | Yes       |
|                      | `http://www.w3.org/2009/xmlenc11#aes256-gcm`          | Yes       |
| Key Transport        | `http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p`     | Yes       |
| Key Transport Digest | `http://www.w3.org/2000/09/xmldsig#sha1`              | Yes       |

# Requirements

- PHP >= 5.4
    - For encryption (`<EncryptedAssertion>`) support, PHP >= 7.1
- `php-openssl`
- See `composer.json` for additional dependencies

# Installation

See [INSTALL](INSTALL.md) for manual installation.

In the near future we'll provide supported CentOS/Fedora and Debian/Ubuntu 
packages.

Currently we have CentOS/Fedora packages in 
[COPR](https://copr.fedorainfracloud.org/coprs/fkooman/php-saml-sp/).

# Source Code Layout

The `src/` directory contains the SAML SP implementation library. The directory
`src/Web` contains everything related to the built-in web interface providing 
the landing page and WAYF. The `src/Api` directory contains everything related
to the API to use from your PHP application.

# Development

Run [composer](https://getcomposer.org/) to install the dependencies:

    $ /path/to/composer install

Use the following command to create self-signed certificates for use with the
SP library. It will be used for signing the `<AuthnRequest>` and 
`<LogoutRequest>`. Another set will be used for decryption of 
`<EncryptedAssertion>`.

    $ cd keys
    $ ./generate_keys.sh

Now copy the configuration template:

    $ cp config/config.php.example config/config.php

Disable the `Secure` session cookie parameter by setting the `secureCookie` key 
to `false` in `config/config.php`.

A neat IdP to use for testing is 
`https://x509idp.moonshot.utr.surfcloud.nl/metadata`. There's no need to register
your SP with that IdP. Put the metadata in `config/metadata`:

    $ mkdir config/metadata
    $ curl -L -o config/metadata/x509idp.moonshot.utr.surfcloud.nl.xml https://x509idp.moonshot.utr.surfcloud.nl/metadata

Run the application using PHP's built-in web server:

    $ php -S localhost:8082 -t web

With your browser you can go to 
[http://localhost:8082/](http://localhost:8082) and take it from there!

# IdP Configuration

In case you want to add / configure your IdP to use with this software, make 
sure:

- the IdP uses the HTTP-Redirect binding for receiving the `AuthnRequest`;
- the IdP uses the HTTP-POST binding to provide the `samlp:Response` to the SP;
- the IdP signs the `saml:Assertion` and/or the `samlp:Response`;
- the IdP (optionally) encrypts the `saml:Assertion`;
- the IdP verifies the signature on the `samlp:AuthnRequest`;
- the IdP verifies the signature on the `samlp:LogoutRequest`;
- the IdP signs the `samlp:LogoutResponse`.

Some of these requirements are also exposed through the SP metadata.

## simpleSAMLphp

In your simpleSAMLphp's `metadata/saml20-sp-remote.php` file, configure this 
for this SP:

    'validate.authnrequest' => true,
    'sign.logout' => true,
    'validate.logout' => true,

# API 

See [API](API.md) for how to use php-saml-sp from your PHP application.

# Tests

In order to run the included test suite:

    $ vendor/bin/phpunit

# Resources

* https://www.owasp.org/index.php/SAML_Security_Cheat_Sheet
* https://www.usenix.org/system/files/conference/usenixsecurity12/sec12-final91-8-23-12.pdf
* https://arxiv.org/pdf/1401.7483v1.pdf
* https://www.cs.auckland.ac.nz/~pgut001/pubs/xmlsec.txt


CHANGES.md
# ChangeLog

## 1.0.2 (2021-01-14)
- handle new PHP 7.3 and 7.4 releases changing `parse_url` behavior 
  (PHP #77423) which broke the PHPUnit tests

## 1.0.1 (2020-12-28)
- fix CSS outline on last chosen IdP in Firefox (issue #12)
- add test for Azure IdP and fix default namespace prefix (issue #11)

## 1.0.0 (2020-12-07)
- fix security issues found by Cure53 audit:
  - DEC-01-001 WP1: ReturnTo validation bypass via URL parser problem (Medium)
  - DEC-01-002 WP1: XSS in example code via unencoded SAML attributes (Medium)
  - DEC-01-007 WP1: CSRF in logout due to missing CSRF protection (Low)
  - DEC-01-004 SP: User can inject XML which is subsequently signed (Low)
  - DEC-01-005 SP: Potential signature bypass via empty string C14N() failure (Low)
  - DEC-01-006 SP: Non-canonicalized XML data used after signature check (Low)
- fix URN of `street` attribute mapping
- improve error message when cookies are disabled in user's browser (issue #3)

## 0.5.9 (2020-10-12)
- `sloUrl` is no longer optional
- make sure `substr` returns the expected number of bytes
- check return value of `DOMDocument::importNode`
- some small `DateTime` type fixes

## 0.5.8 (2020-10-01)
- fix `PublicKey::getFingerprint()` on PHP 5.4
- change mouse cursor to pointer when hovering `<details>`
- add API documentation
- add API method docblocks

## 0.5.7 (2020-08-31)
- implement `http://www.w3.org/2009/xmlenc11#aes128-gcm`, as per saml2int
- add manual install instructions

## 0.5.6 (2020-08-26)
- `php-saml-sp-validate-metadata` shows fingerprints of used certificates. Some
  federations publish those fingerprints, makes it easy to compare
- dynamic metadata MUST still be configured in `config.php`, otherwise it is
  ignored and IdPs from it are no longer available/listed
- update style and remove bootstrap-reboot CSS

## 0.5.5 (2020-08-19)
- fix automatic metadata refresh considering `cacheDuration` / `Last-Modified`

## 0.5.4 (2020-08-12)
- implement CSS/JS cache busting

## 0.5.3 (2020-08-09)
- restore use of `paragonie/constant_time_encoding` for now

## 0.5.2 (2020-07-28)
- show warning on index page when `ext-sodium` is not available
- fix typo in warning for missing encryption support
- update translation
- update `fkooman/secookie`

## 0.5.1 (2020-07-22)
- add the `--force` flag to `php-saml-sp-update-metadata` to be able to force
  metadata retrieval no matter `cacheDuration` in the existing metadata
- make method `MetadataSource::importMetadata` private
- drop `paragonie/constant_time_encoding` dependency and instead use 
  `ext-sodium` functions for constant time encoding
- implement metadata `cacheDuration` support for dynamic metadata to check for 
  new metadata periodically (default is `PT6H`)
- `validUntil` is now only used to reject dynamic metadata, not for 
  determining when to refresh metadata
- follow metadata request redirects
- make sure response code for metadata downloads is `200` before processing
  metadata

## 0.5.0 (2020-07-17)
- implement auto metadata refresh with `metadataList` configuration option 
  specifying the URL and certificate(s) to validate it (issue #8)
- support validating XML metadata signature
- only validate XML schema when importing it, not every time when using it
- metadata files in config/metadata are always trusted (no schema check, no 
  signature validation)
- (**SECURITY**) make sure IdP is allowed to authenticate, previously it was 
  possible to authenticate with an IdP in the metadata, but not explicitly 
  listed under `idpList`
- make it possible to accept _all_ IdPs from metadata without explicitly 
  allowing them by not setting the `idpList` configuration option. An 
  `idpList` with no items, i.e. empty array, does allow NO IdPs
- implement rudimentary logging to _syslog_
- add script to verify metadata XML schema and signature 
  (`php-saml-sp-validate-metadata`)
- ignore non-RSA keys in IdP metadata
- much more helpful error messages in case something is missing/going wrong 
  when parsing XML documents (issue #7)
- major rewrite of XML document handling, no longer expose `DOMDocument` and 
  `DOMXPath`, introduce more helper methods with strict type checks
- remove all XML XPath "evaluate" queries
- make XML signature validation / decryption more robust
- add new line when fetching SP metadata (issue #2)
- better check for empty strings by first calling `trim()` on the string value
- lots of fixes with the help of `vimeo/psalm`, `phpstan/phpstan` and 
  `phan/phan`

## 0.4.2 (2020-07-06)
- support validating ADFS metadata (ws-federation)

## 0.4.1 (2020-06-29)
- implement `<samlp:Scoping>` for `<samlp:AuthnRequest>` to tell a SAML proxy
  which IdP behind the proxy should be used for authentication, thus skipping 
  the WAYF
- support eduMember `isMemberOf` attribute

## 0.4.0 (2020-05-25)
- use `/login` endpoint instead of `/wayf` (issue #4)
- support `ReturnTo` as a query parameter on `/login` (issue #5)
- support `AuthnContextClassRef` as a query parameter on `/login`
- support `IdP` as a query parameter on `/login`
- add (optional) `AuthOptions` parameter to the `SamlAuth` methods for getting
  and verifying the assertion instead of having to do this manually

## 0.3.3 (2020-05-23)
- update `nl-NL` translation files

## 0.3.2 (2020-05-23)
- implement JS search for WAYF (only if browser has JS enabled)
- update bootstrap-reboot CSS
- cleanup language handling / template

## 0.3.1 (2020-02-23)
- use `autofocus` element on the previously selected IdP in the WAYF, so enter
  can be used to continue to that IdP immediately
- translate language code to human readable language for display in UI

## 0.3.0 (2020-02-12)
- small API update for `XmlIdpInfoSource` now taking array in contructor to
  allow specifying multiple metadata files 
- initial support for "Web", i.e. run fkooman/saml-sp as an app on your server
  next to the application(s) you want to use SAML authentication with
- support `<EncryptedAssertion>` with AES-256-GCM on PHP >= 7.1

src/Crypto.php
<?php

/*
 * Copyright (c) 2019-2020 François Kooman <fkooman@tuxed.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

namespace fkooman\SAML\SP;

use fkooman\SAML\SP\Exception\CryptoException;
use ParagonIE\ConstantTime\Base64;

class Crypto
{
    const SIGN_OPENSSL_ALGO = OPENSSL_ALGO_SHA256;
    const SIGN_ALGO = 'http://www.w3.org/2001/04/xmldsig-more#rsa-sha256';
    const SIGN_DIGEST_ALGO = 'http://www.w3.org/2001/04/xmlenc#sha256';
    const SIGN_HASH_ALGO = 'sha256';

    const ENCRYPT_KEY_DIGEST_ALGO = 'http://www.w3.org/2000/09/xmldsig#sha1';
    const ENCRYPT_KEY_TRANSPORT_ALGO = 'http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p';

    /**
     * @param array<PublicKey> $publicKeys
     *
     * @throws \fkooman\SAML\SP\Exception\CryptoException
     *
     * @return void
     */
    public static function verifyXml(XmlDocument $xmlDocument, array $publicKeys)
    {
        $rootElementId = $xmlDocument->requireOneDomAttrValue('self::node()/@ID');
        $referenceUri = $xmlDocument->requireOneDomAttrValue('self::node()/ds:Signature/ds:SignedInfo/ds:Reference/@URI');
        if (\sprintf('#%s', $rootElementId) !== $referenceUri) {
            throw new CryptoException('reference URI does not point to document ID');
        }

        $digestMethod = $xmlDocument->requireOneDomAttrValue('self::node()/ds:Signature/ds:SignedInfo/ds:Reference/ds:DigestMethod/@Algorithm');
        if (self::SIGN_DIGEST_ALGO !== $digestMethod) {
            throw new CryptoException(\sprintf('only digest method "%s" is supported', self::SIGN_DIGEST_ALGO));
        }

        $signatureMethod = $xmlDocument->requireOneDomAttrValue('self::node()/ds:Signature/ds:SignedInfo/ds:SignatureMethod/@Algorithm');
        if (self::SIGN_ALGO !== $signatureMethod) {
            throw new CryptoException(\sprintf('only signature method "%s" is supported', self::SIGN_ALGO));
        }

        $signatureValue = self::removeWhiteSpace($xmlDocument->requireOneDomElementTextContent('self::node()/ds:Signature/ds:SignatureValue'));
        $digestValue = self::removeWhiteSpace($xmlDocument->requireOneDomElementTextContent('self::node()/ds:Signature/ds:SignedInfo/ds:Reference/ds:DigestValue'));

        $signedInfoElement = $xmlDocument->requireOneDomElement('self::node()/ds:Signature/ds:SignedInfo');

        $canonicalSignedInfo = XmlDocument::canonicalizeElement($signedInfoElement);
        $signatureElement = $xmlDocument->requireOneDomElement('self::node()/ds:Signature');
        $rootElement = $xmlDocument->requireOneDomElement('self::node()');
        $rootElement->removeChild($signatureElement);
        $rootElementDigest = Base64::encode(
            \hash(
                self::SIGN_HASH_ALGO,
                XmlDocument::canonicalizeElement($rootElement),
                true
            )
        );

        // compare the digest from the XML with the actual digest
        if (!\hash_equals($rootElementDigest, $digestValue)) {
            throw new CryptoException('unexpected digest');
        }

        self::verify($canonicalSignedInfo, Base64::decode($signatureValue), $publicKeys);
    }

    /**
     * @param string           $inStr
     * @param string           $inSig
     * @param array<PublicKey> $publicKeys
     *
     * @throws \fkooman\SAML\SP\Exception\CryptoException
     *
     * @return void
     */
    public static function verify($inStr, $inSig, array $publicKeys)
    {
        if (0 === \count($publicKeys)) {
            throw new CryptoException('no public key(s) provided');
        }

        // Our XML handling requires the use of a DOMDocument (contained in
        // XmlDocument) that just has the saml:Assertion. This used to work
        // great until we tested with an Azure IdP. It uses the default
        // namespace for the Signature element. Other implementations always
        // use the ds: prefix. This triggers a bug in PHP/libxml where when
        // importing this node into another document, the default namespace is
        // made explicit with the default: prefix. This is very annoying, as
        // this breaks XML signatures...
        //
        // A better solution is probably to NOT import the Assertion in another
        // document and instead do the signature verification directly on the
        // Response document. This works for Assertion, but not
        // EncryptedAssertion. In the latter case the Response XML should be
        // modified and the EncryptedAssertion be replaced by the Assertion.
        // Unfortunately this requires a substantial refactor that I'd like to
        // avoid just after the audit. So instead, we just do simple text
        // replacement to remove the "default" namespace prefix after which the
        // signature verification works again. This is rather safe as if the
        // document is mangled somehow the verification will simply fail. As no
        // other part of the Signature element can possibly contain these
        // strings I'm rather sure this is fine...
        //
        // @see https://bugs.php.net/bug.php?id=55294
        // @see https://bugs.php.net/bug.php?id=47530
        $inStr = \str_replace(['<default:', '</default:', 'xmlns:default='], ['<', '</', 'xmlns='], $inStr);

        foreach ($publicKeys as $publicKey) {
            if (1 === \openssl_verify($inStr, $inSig, $publicKey->raw(), self::SIGN_OPENSSL_ALGO)) {
                // signature verified
                return;
            }
        }

        throw new CryptoException('unable to verify signature');
    }

    /**
     * @throws \fkooman\SAML\SP\Exception\CryptoException
     *
     * @return string
     */
    public static function decryptXml(XmlDocument $xmlDocument, PrivateKey $privateKey)
    {
        if (!self::hasDecryptionSupport()) {
            throw new CryptoException('XML decryption is not supported on this system');
        }

        // make sure we support the key transport encryption algorithm
        $keyEncryptionMethod = $xmlDocument->requireOneDomAttrValue('/samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData/ds:KeyInfo/xenc:EncryptedKey/xenc:EncryptionMethod/@Algorithm');
        if (self::ENCRYPT_KEY_TRANSPORT_ALGO !== $keyEncryptionMethod) {
            throw new CryptoException(\sprintf('only key transport algorithm "%s" is supported', self::ENCRYPT_KEY_TRANSPORT_ALGO));
        }

        // make sure we support the key transport encryption digest algorithm
        $keyEncryptionDigestMethod = $xmlDocument->requireOneDomAttrValue('/samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData/ds:KeyInfo/xenc:EncryptedKey/xenc:EncryptionMethod/ds:DigestMethod/@Algorithm');
        if (self::ENCRYPT_KEY_DIGEST_ALGO !== $keyEncryptionDigestMethod) {
            throw new CryptoException(\sprintf('only key encryption digest "%s" is supported', self::ENCRYPT_KEY_DIGEST_ALGO));
        }

        // extract the session key
        $keyCipherValue = self::removeWhiteSpace($xmlDocument->requireOneDomElementTextContent('/samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData/ds:KeyInfo/xenc:EncryptedKey/xenc:CipherData/xenc:CipherValue'));

        // decrypt the session key
        if (false === \openssl_private_decrypt(Base64::decode($keyCipherValue), $symmetricEncryptionKey, $privateKey->raw(), OPENSSL_PKCS1_OAEP_PADDING)) {
            throw new CryptoException('unable to extract session key');
        }

        // find out which encryption algorithm was used for the EncryptedAssertion
        $encryptionMethod = $xmlDocument->requireOneDomAttrValue('/samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData/xenc:EncryptionMethod/@Algorithm');

        // make sure the obtained key is the exact length we expect
        if (CryptoParameters::getKeyLength($encryptionMethod) !== \strlen($symmetricEncryptionKey)) {
            throw new CryptoException('session key has unexpected length');
        }

        // extract the encrypted Assertion
        $assertionCipherValue = Base64::decode(self::removeWhiteSpace($xmlDocument->requireOneDomElementTextContent('/samlp:Response/saml:EncryptedAssertion/xenc:EncryptedData/xenc:CipherData/xenc:CipherValue')));

        // @see https://www.w3.org/TR/xmlenc-core1/#sec-AES-GCM

        // IV (first 96 bits)
        $messageIv = self::expectStringWithLen(
            \substr($assertionCipherValue, 0, CryptoParameters::getIvLength($encryptionMethod)),
            CryptoParameters::getIvLength($encryptionMethod)
        );
        // Tag (last 128 bits)
        $messageTag = self::expectStringWithLen(
            \substr($assertionCipherValue, \strlen($assertionCipherValue) - CryptoParameters::getTagLength($encryptionMethod)),
            CryptoParameters::getTagLength($encryptionMethod)
        );
        // CipherText (between IV and Tag)
        $cipherText = self::expectStringWithLen(
            \substr($assertionCipherValue, CryptoParameters::getIvLength($encryptionMethod), -CryptoParameters::getTagLength($encryptionMethod)),
            \strlen($assertionCipherValue) - CryptoParameters::getIvLength($encryptionMethod) - CryptoParameters::getTagLength($encryptionMethod)
        );
        if (false === $decryptedAssertion = \openssl_decrypt($cipherText, CryptoParameters::getOpenSslAlgorithm($encryptionMethod), $symmetricEncryptionKey, OPENSSL_RAW_DATA, $messageIv, $messageTag)) {
            throw new CryptoException('unable to decrypt data');
        }

        return $decryptedAssertion;
    }

    /**
     * @param string $inStr
     *
     * @throws \fkooman\SAML\SP\Exception\CryptoException
     *
     * @return string
     */
    public static function sign($inStr, PrivateKey $privateKey)
    {
        if (false === \openssl_sign($inStr, $outSig, $privateKey->raw(), self::SIGN_OPENSSL_ALGO)) {
            throw new CryptoException('unable to create signature');
        }

        return $outSig;
    }

    /**
     * @return bool
     */
    public static function hasDecryptionSupport()
    {
        // ext-openssl MUST be loaded
        if (false === \extension_loaded('openssl')) {
            return false;
        }

        // we need PHP >= 7.1.0 for AEAD support in PHP/OpenSSL
        if (\version_compare(PHP_VERSION, '7.1.0') < 0) {
            return false;
        }

        // we need support for AES-GCM
        foreach (['aes-128-gcm', 'aes-256-gcm'] as $encryptionAlgorithm) {
            if (!\in_array($encryptionAlgorithm, \openssl_get_cipher_methods(), true)) {
                return false;
            }
        }

        return true;
    }

    /**
     * @param string $inputStr
     *
     * @return string
     */
    private static function removeWhiteSpace($inputStr)
    {
        // https://www.w3.org/TR/xmlschema11-2/#base64Binary
        // https://www.w3.org/TR/xmlschema11-2/#rf-whiteSpace
        // we do NOT follow the exact rules for white space removal, because
        // they are insane...
        return \str_replace(["\x09", "\x0a", "\x0d", "\x20"], '', $inputStr);
    }

    /**
     * @param mixed $inputData
     * @param int   $expectedLength
     *
     * @return string
     */
    private static function expectStringWithLen($inputData, $expectedLength)
    {
        if (!\is_string($inputData)) {
            throw new CryptoException('"string" expected');
        }
        if ($expectedLength !== \strlen($inputData)) {
            throw new CryptoException('unexpected "string" length');
        }

        return $inputData;
    }
}
